src/pointer-analysis/dereference.cpp

```diff
@@ -612,6 +612,64 @@ bool dereferencet::dereference_type_compare(
   return false;
 }
 
+void dereferencet::check_pointer_alignment(
+  modet mode,
+  const type2tc &type,
+  const expr2tc &deref_expr,
+  const guardt &guard)
+{
+  if (options.get_bool_option("no-align-check"))
+    return;
+
+  // Only check alignment for scalar read/write operations (excluding code and pointer types)
+  if (
+    !(is_read(mode) || is_write(mode)) || !is_scalar_type(type) ||
+    is_code_type(type) || is_pointer_type(type))
+  {
+    return;
+  }
+
+  BigInt access_size_bits = type_byte_size_bits(type);
+
+  // Only check alignment for byte-aligned accesses
+  if (access_size_bits % 8 != 0)
+    return;
+
+  expr2tc ptr_offset_bits = create_pointer_offset_bits(deref_expr);
+  simplify(ptr_offset_bits);
+  check_alignment(access_size_bits, ptr_offset_bits, guard);
+}
+
+expr2tc dereferencet::create_pointer_offset_bits(const expr2tc &deref_expr)
+{
+  // Handle constant integer expressions
+  if (is_constant_int2t(deref_expr))
+  {
+    return constant_int2tc(
+      bitsize_type2(), to_constant_int2t(deref_expr).value * 8);
+  }
+
+  // Handle typecast expressions with constant integer sources
+  if (
+    is_typecast2t(deref_expr) &&
+    is_constant_int2t(to_typecast2t(deref_expr).from))
+  {
+    return constant_int2tc(
+      bitsize_type2(),
+      to_constant_int2t(to_typecast2t(deref_expr).from).value * 8);
```

Comment:
This smells a bit unsound-ish to me.
Wouldn't `create_pointer_offset_bits` return 8000 for `(int8_t) (256*4 - 24)` (256*4 - 24 = 1000) while instead it should return `-192`?

Reply:
Well spotted, @intrigus-lgtm.

I have just encoded the pointer offset calculation symbolically and let the `check_alignment` (https://github.com/esbmc/esbmc/blob/master/src/pointer-analysis/dereference.cpp#L2415-L2460) do the computation: https://github.com/esbmc/esbmc/commit/ff23dd763608f13b9575ca18a2cfbdf94910677e. 

This has found a regression that was incorrectly classified in our suite: https://github.com/esbmc/esbmc/commit/0daf88c840a162e97d163e2c04fd6585345037fa.

Reply:
> This has found a regression that was incorrectly classified in our suite: [0daf88c](https://github.com/esbmc/esbmc/commit/0daf88c840a162e97d163e2c04fd6585345037fa).

I'm still not 100% sold on what we are doing here.
That is, `((b *)60000)->a = 0;` rightfully crashes in most environments such as the "normal" linux environment.
But if you consider memory-mapped IO instead, this code might be perfectly valid, isn't it?

Reply:
That's a good question.

@rafaelsamenezes has worked on this with one of our industrial partners.

@rafaelsamenezes: Do you have any thoughts here?